POC of minimal events to support indexer and RPC calls (#460)
mangrovedao/mangrove-core: Events.md

# Introduction

Events in solidity is a hard thing to do in a optimal way. If you look at it as a purely gas efficient issue, you want to emit as few events as possible and with as few fields as possible. But as events also has to be usable for an off chain user, then doing this is not always the best solution.

We tried to list the main points that we would like to do with events.

1. Use as little gas as possible
2. An indexer should be able to keep track of the state of Mangrove.
3. Doing RPC calls directly, should be able to find offers and other information based on offerlist, maker, taker, offer id, etc.

These 3 points all have there own direction and it is therefore not possible to find a solution that is optimal for all 3 points.
We have therefore tried to find a solution that is a good balance between the all 3 points.

> ## Events

Reviewer: IMHO all the event-specific documentation should be inlined where the events are defined. @adhusson wdyt?

Reviewer 2: All the documentation has been inlined at their relevant event.

Approvals are based on the pair of outbound and inbound token. Be aware that it is not offerlist bases, as an offerlist also holds the tickscale.

We emit `outbound` token, `inbound` token, `owner`, msg.sender (`spender`), `value`. Where `owner` is the one who owns the funds, `spender` is the one who is allowed to use the funds and `value` is the amount of funds that is allowed to be used.

Outbound, inbound and owner is indexed, this way one can filter on these fields when doing RPC calls. If we could somehow combine outbound and inbound into one field, then we could also index the spender.

### Credit

This is emitted when a user's account on Mangrove is credited with some native funds, to be used as provision for offers.

It emits the `maker`'s address and the `value` credited

The `maker` address is indexed so that we can filter on it when doing RPC calls.

These are the scenarios where it can happen:

- Fund Mangrove directly
- Fund Mangrove when posting an offer
- When updating an offer
  - Funding Mangrove or
  - the updated offer needs less provision, than it already has. Meaning the user gets credited the difference.

> - When retracting an offer and deprovisioning it.Meaning the user gets credited the provision that was locked by the offer.

Reviewer: ```suggestion
- When retracting an offer and deprovisioning it. Meaning the user gets credited the provision that was locked by the offer.
```

- When an offer fails. The remaining provision gets credited back to the maker

A challenge for an indexer is to know how much provision each offer locks. With the current events, an indexer is going to have to know the liveness, gasreq and gasprice of the offer. If the offer is not live, then also know if it has been deprovisioned. And know what the gasbase of the offerlist was when the offer was posted. With this information an indexer can calculate the exact provision locked by the offer.

> Another information that an indexer cannot deduce, is in what scenario the credit event happened. E.g. We don't know if the credit event happened because an offer failed or if the user simply funded Mangrove.

Reviewer: ```suggestion
The indexer also cannot deduce what scenario the credit event happened. E.g., we don't know if the credit event happened because an offer failed or because the user simply funded Mangrove.
```

### Debit

This is emitted when a user's account on Mangrove is debited with some native funds.

It emits the `maker`'s address and the `value` debited.

The `maker` address is indexed so that we can filter on it when doing RPC calls.

These are the scenarios where it can happen:

- Withdraw funds from Mangrove directly
- When posting an offer. The user gets debited the provision that the offer locks.
- When updating an offer and it requires more provision that it already has. Meaning the user gets debited the difference.

Same challenges as for Credit

### Kill

Emitted when the Mangrove instance is killed

### NewMgv

Emitted when a new Mangrove is deployed.

### OrderStart

This event is emitted when a market order is started on Mangrove.

> It emits the `offerlist key`, the `taker`, the `maxLogPrice`, `fillVolume` and `fillWants`.

Reviewer: offerlist => offerList in general - and it is the hash that gets emitted, right?

Reviewer 2: Fixed

The fields `offerlist key` and `taker` are indexed, so that we can filter on them when doing RPC calls.

> By emitting this an indexer can keep track of what context the current market order is in. E.g. If a user starts a market order and one of the offers taken also starts a market order, then we can in an indexer have a stack of started market orders and thereby know exactly what offerlist the order is running on and the taker.

Reviewer: In general, E.g. should not be followed by capitalized letter.

Reviewer 2: Fixed

By emitting `maxLogPrice`, `fillVolume` and `fillWants`, we can now also know how much of the market order was filled and if it matches the price given. See OrderComplete for more.

### OfferFail(WithPosthookData)

This event is emitted when an offer fails, because of a maker error.

It emits `offerlist key`, `taker`, `maker`, the `offerId`, the offers `wants`, `gives`, `penalty` and the `reason` for failure. `Offer list key`, `taker` and `maker` are all fields that we do not need, in order for an indexer to work, as an indexer will be able the get that info from the former `OrderStart` and `OfferWrite` events. But in order for RPC call to filter on this, we need to emit them. Those 3 fields are indexed for the same reason.

If the posthook of the offer fails. Then we emit `OfferFailWithPosthookData` instead of just `OfferFail`. This event has one extra field, which is the reason for the posthook failure. By emitting the posthook data, an indexer can keep track of the reason posthook fails, this could for example be used for analytics.

> This event is emitted doring posthook end, we wait to emit this event to the end, because we need the information of `penalty`, which is only available at the end of the posthook. This means that `OfferFail` events are emitted in reverse order, compared to what order they are taken. This is do to the way we handle posthooks. The same goes for `OfferSuccess`.

Reviewer: ```suggestion
This event is emitted doring posthook end, we wait to emit this event to the end, because we need the information of `penalty`, which is only available at the end of the posthook. This means that `OfferFail` events are emitted in reverse order, compared to what order they are taken. This is due to the way we handle posthooks. The same goes for `OfferSuccess`.
```

If the posthook of the offer fails. Then we emit `OfferSuccessWithPosthookData` instead of just `OfferSuccess`. This event has one extra field, which is the reason for the posthook failure. By emitting the posthook data, an indexer can keep track of the reason posthook fails, this could for example be used for analytics.

By emitting `offerId`, `wants` and `gives`, an indexer can keep track of whether the offer was partially or fully taken, by looking at the what the offer was posted at.

### OrderComplete

This event is emitted when a market order is finished.

It only emits the total `fee paid`. We need this event, in order to know that a market order is completed. This way an indexer can know exactly in what context we are in.

> The total `fee paid` for that market order, is needed, as we do not emit this any other place. It is not indexed as there is no need for RPC calls to filter on this.

Author: Perhaps this event should have the same indexed fields as `OrderStart` so RPC clients can filter on this and get the complete picture for the order?

Reviewer 2: Might not be a bad idea. @adhusson @lnist   WDYT?

Reviewer: I agree

Reviewer 2: I've implemented this now.

The fields `offerlist key` and `maker` are indexed, so that we can filter on them when doing RPC calls.

### CleanStart

This event is emitted when a user tries to clean offers on Mangrove, using the build in clean functionality.

It emits the `offerlist key`, the `taker` address and `offersToBeCleaned`, which is the number of offers that should be cleaned. By emitting this event, an indexer can save what `offerlist` the user is trying to clean and what `taker` is being used. This way it can keep a context for the following events being emitted (Just like `OrderStart`). The `offersToBeCleaned` is emitted so that an indexer can keep track of how many offers the user tried to clean. Combining this with the amount of `OfferFail` events emitted, then an indexer can know how many offers the user actually managed to clean. This could be used for analytics.

The fields `offerlist key` and `taker` are indexed, so that we can filter on them when doing RPC calls.

### CleanComplete

This event is emitted when a Clean operation is completed.

> It does not emit any fields. This event is only needed in order to know that the clean operation is completed. This way an indexer can know exactly in what context we are in. It could emit the total bounty received, but in order to know who got the bounty, an indexer would still be needed or we would have to emit the taker address aswell, in order for an RPC call to find the data. But an indexer would still be able to find this info, buy collecting all the previous events. So we do not emit the bounty.

Reviewer: ```suggestion
It does not emit any fields. This event is only needed in order to know that the clean operation is completed. This way an indexer can know exactly in what context we are in. It could emit the total bounty received, but in order to know who got the bounty, an indexer would still be needed or we would have to emit the taker address as well, in order for an RPC call to find the data. But an indexer would still be able to find this info, by collecting all the previous events. So we do not emit the bounty.
```
